artemis-commons/src/main/java/org/apache/activemq/artemis/utils/LazyHashProcessor.java

```diff
@@ -0,0 +1,63 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.activemq.artemis.utils;
+
+public abstract class LazyHashProcessor implements HashProcessor {
+   private volatile HashProcessor hashProcessor;
+   private volatile RuntimeException supplierException;
+
+   public HashProcessor getHashProcessor() {
+      if (supplierException != null) {
+         throw supplierException;
+      }
+
+      if (hashProcessor == null) {
+         synchronized (this) {
+            if (supplierException != null) {
+               throw supplierException;
+            }
+
+            if (hashProcessor == null) {
+               try {
+                  hashProcessor = createHashProcessor();
+               } catch (RuntimeException e) {
+                  supplierException = e;
+                  throw supplierException;
+               } catch (Exception e) {
+                  supplierException = new RuntimeException(e);
+                  throw supplierException;
+               }
+            }
+         }
+      }
+
+      return hashProcessor;
+   }
```

Comment:
```suggestion
public abstract class LazyHashProcessor implements HashProcessor {
   private volatile HashProcessor hashProcessor;
   private RuntimeException supplierException;

   public HashProcessor getHashProcessor() {
      HashProcessor hp = hashProcessor;
      if (hp == null) {
         synchronized (this) {
            if (supplierException != null) {
               throw supplierException;
            }

            hp = hashProcessor;
            if (hp == null) {
               try {
                  hashProcessor = hp = createHashProcessor();
               } catch (RuntimeException e) {
                  supplierException = e;
                  throw supplierException;
               } catch (Exception e) {
                  supplierException = new RuntimeException(e);
                  throw supplierException;
               }
            }
         }
      }

      return hp;
   }
```
I think the initial if (supplierException != null) bit can just be removed, and that field doesnt then need to be volatile as its then only used under synchronization. The exception will only be non-null if hashProcessor is still null, and whole the idea is to typically not expect it to be null, so it seems unnecessary to pre-check the exception every time through. It can just rely on the check at the start of the synchronized block.

I think it should also use the local variable to copy hashProcessor to as typically expected in these cases. 

Something like the suggestion (didnt check it compiles).

Reply:
@franz1981 @gemmellr thanks I replaced the getHashProcessor with the MemoizingSupplier.